plugins/dell-kestrel/fu-dell-kestrel-hid-device.c

```diff
@@ -0,0 +1,401 @@
+		return FALSE;
+
+	if (delayms > 0)
+		fu_device_sleep(FU_DEVICE(self), delayms);
+
+	if (!fu_dell_kestrel_hid_device_get_report(self, inbuf, error))
+		return FALSE;
+
+	return fu_memcpy_safe(res->data, res->len, 0, inbuf, sizeof(inbuf), 1, res->len, error);
+}
+
+static guint
+fu_dell_kestrel_hid_device_get_chunk_delaytime(FuDellKestrelEcDevType dev_type)
+{
+	switch (dev_type) {
+	case FU_DELL_KESTREL_EC_DEV_TYPE_MAIN_EC:
+		return 3 * 1000;
+	case FU_DELL_KESTREL_EC_DEV_TYPE_RMM:
+		return 60 * 1000;
+	case FU_DELL_KESTREL_EC_DEV_TYPE_PD:
+		return 15 * 1000;
+	case FU_DELL_KESTREL_EC_DEV_TYPE_LAN:
+		return 70 * 1000;
+	default:
+		return 30 * 1000;
+	}
+}
+
+static gsize
+fu_dell_kestrel_hid_device_get_chunk_size(FuDellKestrelEcDevType dev_type)
+{
+	/* return the max chunk size in bytes */
+	switch (dev_type) {
+	case FU_DELL_KESTREL_EC_DEV_TYPE_MAIN_EC:
+		return FU_DELL_KESTREL_HID_DEV_EC_CHUNK_SZ;
+	case FU_DELL_KESTREL_EC_DEV_TYPE_RMM:
+		return FU_DELL_KESTREL_HID_DEV_NO_CHUNK_SZ;
+	default:
+		return FU_DELL_KESTREL_HID_DEV_ANY_CHUNK_SZ;
+	}
+}
+
+static guint
+fu_dell_kestrel_hid_device_get_first_page_delaytime(FuDellKestrelEcDevType dev_type)
+{
+	return (dev_type == FU_DELL_KESTREL_EC_DEV_TYPE_RMM) ? 75 * 1000 : 0;
+}
+
+static gboolean
+fu_dell_kestrel_hid_device_write_firmware_pages(FuDellKestrelHidDevice *self,
+						FuChunkArray *pages,
+						FuProgress *progress,
+						FuDellKestrelEcDevType dev_type,
+						guint chunk_idx,
+						GError **error)
+{
+	guint first_page_delay = fu_dell_kestrel_hid_device_get_first_page_delaytime(dev_type);
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(pages));
+
+	for (guint j = 0; j < fu_chunk_array_length(pages); j++) {
+		g_autoptr(GByteArray) page_aligned = g_byte_array_new();
+		g_autoptr(FuChunk) page = NULL;
+		g_autoptr(GBytes) page_bytes = NULL;
+		g_autoptr(GError) error_local = NULL;
+
+		page = fu_chunk_array_index(pages, j, error);
+		if (page == NULL)
+			return FALSE;
+
+		g_debug("sending chunk: %u, page: %u/%u.",
+			chunk_idx,
+			j,
+			fu_chunk_array_length(pages) - 1);
+
+		/* strictly align the page size with 0x00 as packet */
+		g_byte_array_append(page_aligned,
+				    fu_chunk_get_data(page),
+				    fu_chunk_get_data_sz(page));
+		fu_byte_array_set_size(page_aligned, FU_DELL_KESTREL_HID_DATA_PAGE_SZ, 0xFF);
+
+		/* send to ec */
+		if (!fu_dell_kestrel_hid_device_write(self, page_aligned, &error_local)) {
+			/* A buggy device may fail to send an acknowledgment receipt
+			   after the last page write, resulting in a timeout error.
+
+			   This is a known issue so waive it for now.
+			*/
+			if (dev_type == FU_DELL_KESTREL_EC_DEV_TYPE_LAN &&
+			    j == fu_chunk_array_length(pages) - 1 &&
+			    g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_TIMED_OUT)) {
+				g_debug("ignored error: %s", error_local->message);
+				continue;
```

Comment:
ohhhh! we leave early so `fu_progress_step_done(progress)` is required to update the progress.

Reply:
Yup, exactly that -- can you add that to this branch please.